### DISCUSSION TITLE
Tycho build maven artifact (org.jboss.forge.roaster.roaster-api)

### DISCUSSION THREAD

**Discussion by antoniogarmendia** [eclipse-tycho/tycho#1007] [Q&A]
Hi all,

I am trying to include this maven plugin: org.jboss.forge.roaster.roaster-api in my deployment, but I am getting this error:

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.3:validate-classpath (default-validate-classpath) on project generate.java.code: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.3:validate-classpath failed: org.osgi.framework.BundleException: Bundle generate.java.code cannot be resolved:generate.java.code [3]
[ERROR]   Unresolved requirement: Require-Bundle: org.jboss.forge.roaster.roaster-api
[ERROR]     -> Bundle-SymbolicName: org.jboss.forge.roaster.roaster-api; bundle-version="2.25.0.Final"; singleton:="true"
[ERROR]        org.jboss.forge.roaster.roaster-api [1]
[ERROR]          Unresolved requirement: Require-Capability: osgi.extender; filter:="(osgi.extender=osgi.serviceloader.processor)"
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.3:validate-classpath (default-validate-classpath) on project generate.java.code: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.3:validate-classpath failed: org.osgi.framework.BundleException: Bundle generate.java.code cannot be resolved:generate.java.code [3]
  Unresolved requirement: Require-Bundle: org.jboss.forge.roaster.roaster-api
    -> Bundle-SymbolicName: org.jboss.forge.roaster.roaster-api; bundle-version="2.25.0.Final"; singleton:="true"
       org.jboss.forge.roaster.roaster-api [1]
         Unresolved requirement: Require-Capability: osgi.extender; filter:="(osgi.extender=osgi.serviceloader.processor)"

Since I am using Apache Aries I do not know if I am missing something. I am working with Tycho 2.7.3.

I was able to implement a minimal example and the same error appear. 

[minimal.example.project.zip](https://github.com/eclipse/tycho/files/8870598/minimal.example.project.zip)

This minimal example is pomless and has the following structure:
- bundles: has only one plugin called generate.java.code
- features: contains the project with the feature project
- launch: run tycho. Basically, "clean verify -X"
- releng: contains the target paltform, the tycho configuration and the update site project

In this example, I added the Apache Aries to the target platform, but I am not sure if this is necessary for compilation.

What I am missing? Any hints?

Thanks in advance.

Kind regards,
Antonio

**Comment by Bananeweizen** [comment]
You might try adding explicit POM requirements for the Aries bundles: https://stackoverflow.com/questions/43775025/after-jetty-update-to-9-4-x-missing-constraint-require-capability-osgi-extend.
I have only used Aries with the Serviceloader (for automatic JUnit 5 extension registration), that worked without explicit POM dependencies, however.

**Comment by antoniogarmendia** [comment]
Hi Michael,

Thanks for your answer. Unfortunately, It seems that it does not work for me. I updated the minimal example and added those three dependencies but I got the same error.

[minimal.example.projectv1.zip](https://github.com/eclipse/tycho/files/8889455/minimal.example.projectv1.zip)

Any other hints?

Thanks.

Kind regards,
Antonio

**Reply by mickaelistria** [reply]
I believe that p2, which is used by Eclipes Tycho to compute dependencies to download, is failing at honoring this requirement and thus the spi-fly bundles are not added as dependencies, resulting in later errors. You may want to customize your build (either in pom.xml target-platform-configuration, or in build.properties additional.bundles) to add those bundles as build requirements.

**Reply by antoniogarmendia** [reply]
Hi all,

Thanks Michael! Indeed, that was another thing that I missed. However, when the plug-in is installed there is a loaded error and I think that the problems is the start levels. The plugin is loaded correctly when:
org.apache.aries.util -> start-level 1
org.apache.aries.spifly.dynamic.bundle-> start-level 1
org.jboss.forge.roaster.roaster-api -> start-level 2
org.jboss.forge.roaster.roaster-jdt -> start-level 2

I think that I basically have this problem [1]. However, I do not know how to add this because these 4 jars are added to the target platform via maven. I made some searches like [2]. It is possible to add the start levels to the target platform? tycho+maven support this?

Thanks.

Kind regards,
Antonio

P.S: I updated the minimal example

[minimal.example.project.zip](https://github.com/eclipse/tycho/files/8908501/minimal.example.project.zip)

[1] https://stackoverflow.com/questions/14025012/defining-auto-start-for-a-plug-in-within-an-eclipse-update-site-project 
[2] https://bnd.bndtools.org/chapters/305-startlevels.html

**Reply by mickaelistria** [reply]
As far as I know, at the moment, adding Maven artifacts to target-platform wouldn't allow you to override p2 metadata, it doesn't allow a p2.inf; so it seems tricky to add auto-start or start-level info in the p2 repo. But if your deployment is a .product, you can configure start levels in it.

**Reply by antoniogarmendia** [reply]
Hi,

I imagine that. The thing is that my deployment is not a .product, so I need to find a solution for this. Do you know any workaround? 

What I was thinking is in implementing the extension point "org.eclipse.ui.startup" and programmatically load aries first and then, roaster. Do you think this is gonna work? It does not seem to me a nice approach, but....

Thanks.   

Kind regards,
Antonio

**Reply by mickaelistria** [reply]
One workaround is to mimic what product do to set startLevel. They create and require a unit that is attached to the bundle unit (sets it as `host`) and use the `setStartLevel` touchpoint.

```xml
    <unit id='toolingcocoa.macosx.aarch64org.eclipse.equinox.common' version='4.24.0.I20220607-0700' singleton='false'>
      <hostRequirements size='2'>
        <required namespace='osgi.bundle' name='org.eclipse.equinox.common' range='0.0.0'/>
        <required namespace='org.eclipse.equinox.p2.eclipse.type' name='bundle' range='[1.0.0,2.0.0)' greedy='false'/>
      </hostRequirements>
      <properties size='1'>
        <property name='org.eclipse.equinox.p2.type.fragment' value='true'/>
      </properties>
      <provides size='2'>
        <provided namespace='org.eclipse.equinox.p2.iu' name='toolingcocoa.macosx.aarch64org.eclipse.equinox.common' version='4.24.0.I20220607-0700'/>
        <provided namespace='org.eclipse.equinox.p2.flavor' name='toolingcocoa.macosx.aarch64' version='1.0.0'/>
      </provides>
      <requires size='2'>
        <required namespace='osgi.bundle' name='org.eclipse.equinox.common' range='0.0.0'/>
        <required namespace='org.eclipse.equinox.p2.eclipse.type' name='bundle' range='[1.0.0,2.0.0)' greedy='false'/>
      </requires>
      <filter>
        (&amp;(osgi.arch=aarch64)(osgi.os=macosx)(osgi.ws=cocoa))
      </filter>
      <touchpoint id='null' version='0.0.0'/>
      <touchpointData size='1'>
        <instructions size='4'>
          <instruction key='uninstall'>
            uninstallBundle(bundle:${artifact})
          </instruction>
          <instruction key='install'>
            installBundle(bundle:${artifact})
          </instruction>
          <instruction key='configure'>
            setStartLevel(startLevel:2);markStarted(started: true);
          </instruction>
          <instruction key='unconfigure'>
            setStartLevel(startLevel:-1);markStarted(started: false);
          </instruction>
        </instructions>
      </touchpointData>
    </unit>
```

**Reply by antoniogarmendia** [reply]
Hi Michael,

I am not quite sure what exactly I should do or look at. I have been reading [1] and [2] but I do not know where this config file you shared belongs to. Do I need to create a fragment project where the fragment host is the actual plugin that I want to put a startup level on?

Would you please give me more insights about this? Any Tutorial? Thanks.

Kind regards,
Antonio

[1] https://wiki.eclipse.org/Installable_Units
[2] https://wiki.eclipse.org/Equinox/p2/Setting_Start_Levels

**Reply by mickaelistria** [reply]
This snippet I shared comes from the content.jar file in https://download.eclipse.org/eclipse/updates/4.24/R-4.24-202206070700/ . This are description of installable units that allow to set start levels. You can create them from a p2.inf file is you use Tycho; creating a whole `eclipse-installable-unit` for those, or adding this configuration for to some feature build.

**Reply by antoniogarmendia** [reply]
Hi Michael,

I think I am near, but still is not working. I take look at [1] and [2]. Since I am using Tycho I think that I can add the p2.inf co-located with the feature.xml. Is this correct? For instance, for apache.aries it looks something like this:

```
# Set start level for org.apache.aries.util
requires.0.namespace = org.eclipse.equinox.p2.iu
requires.0.name = org.apache.aries.util.fragment
requires.0.range = 1.0.0

#Definitions org.apache.aries.util
units.0.id = org.apache.aries.util.fragment
units.0.version = 1.0.0
units.0.singleton = true
units.0.provides.0.namespace = org.eclipse.equinox.p2.iu
units.0.provides.0.name = org.apache.aries.util.fragment
units.0.provides.0.version = 1.0.0
units.0.instructions.configure = setStartLevel(startLevel:1);markStarted(started: true);
units.0.hostRequirements.0.namespace = org.eclipse.equinox.p2.iu
units.0.hostRequirements.0.name = org.apache.aries.util
units.0.hostRequirements.0.version = 0.0.0
units.0.instructions.install = installBundle(bundle:${artifact})
units.0.instructions.uninstall = uninstallBundle(bundle:${artifact})
```

In the content.jar/content.xml I cannot find the touchpointData with the startlevel (I would say that it should be very similar to the code you shared). I updated the minimal example just in case you have the time to take a look.

[minimal.example.project.zip](https://github.com/eclipse/tycho/files/8942253/minimal.example.project.zip)

Do you have a hint of what I am missing? Thanks.

Kind regards,
Antonio

[1] https://wiki.eclipse.org/Equinox/p2/Customizing_Metadata#Additional_Installable_Unit_Advice:
[2] https://www.eclipse.org/forums/index.php/t/152692/

**Reply by mickaelistria** [reply]
Is the p2.inf added to build.properties so it is available in the .jar file? The way publishing is done currently mandates that (it's a limitation that was discussed recently in #1014 )

**Reply by antoniogarmendia** [reply]
Hi Michael,

Indeed, it is added.

```
bin.includes = feature.xml,\
               p2.inf
```

Just if you have the time here is the updated minimal example.

[minimal.example.project.zip](https://github.com/eclipse/tycho/files/8948560/minimal.example.project.zip)

The compilation succeeded, but when I try to install this plugin it gives me this error.

```
Cannot complete the install because one or more required items could not be found.
  Software being installed: Feature 1.0.0.202206211145 (minimal.example.feature.feature.group 1.0.0.202206211145)
  Missing requirement: Feature 1.0.0.202206211145 (minimal.example.feature.feature.group 1.0.0.202206211145) requires 'org.eclipse.equinox.p2.iu; org.apache.aries.util.fragment 1.0.0' but it could not be found
```

Is kind of weird because the org.apache.aries.util.fragment is defined in the p2.inf and the plugin org.apache.aries.util is an included plugin.

Any other hint of what I am missing? Thanks.

Kind regards,
Antonio

**Reply by mickaelistria** [reply]
I looked at the example, and it's not really minimized; it requires xtend, has many deps... A standalone Tycho-only example would be easier for me to try.

**Reply by antoniogarmendia** [reply]
Hi Michael,

Sorry, I removed Xtend dependency (It is really not necessary:-/) and reduced the pom as much as I could, but the other dependencies are needed in order to roaster compiles correctly. It has 5 dependencies, but all of them are added in the target platform.

This is the updated minimal example: 

[minimal.example.project.zip](https://github.com/eclipse/tycho/files/8949419/minimal.example.project.zip)

This minimal example is pomless and has the following structure:

   - bundles: has only one plugin called generate.java.code
   - features: contains the project with the feature project (minimal.example.feature)
   - launch called "minimal.example.project": run tycho. Basically, "clean verify -X"
   - releng: contains the target paltform (minimal.example.target\minimal.example.target.), the tycho configuration and the update site project.

Do you need more info? Do I need to reduce this even more? Please, do not hesitate to let me know.

Thanks!

Kind regards,
Antonio

**Reply by mickaelistria** [reply]
I gave it a try. The unit is properly generated in the feature as one can see in features/minimal.example.feature/target/p2content.xml , but the unit is not included into the p2 repo. To force inclusion in the p2 repo, you can add `<iu id="org.apache.aries.util.fragment"/>` in the category.xml, just like you'd add a bundle.

**Reply by laeubi** [reply]
Why not just include the fragment in the feature if its required?

**Reply by antoniogarmendia** [reply]
Hi all,

@mickaelistria Thanks for giving it a try. Indeed, I missed that on the category.xml (I just read [1]). However, It does not work for me when I install this plugin. I extended the previous minimal example with other units that I also need. In my runtime Eclipse I have the following start-level configuration:

aries.util->startlevel:1
aries.spifly.dynamic.bundle->startlevel:1
roaster-api->startlevel:2
roaster-jdt->startlevel:2

In the content.jar/content.xml I can see that the start levels are well defined. However, when I install these fragments I got the following error:

```
org.osgi.framework.BundleException: Could not resolve module: org.jboss.forge.roaster.roaster-jdt [1628]
  Unresolved requirement: Import-Package: com.ibm.cds; resolution:="optional"
  Unresolved requirement: Import-Package: com.ibm.oti.shared; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster._shade.org.eclipse.core.filesystem; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster._shade.org.eclipse.core.filesystem.provider; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster._shade.org.eclipse.core.internal.runtime.auth; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster._shade.org.eclipse.equinox.app; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster._shade.org.eclipse.equinox.internal.app; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster._shade.org.eclipse.osgi.internal.resolver; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster.model; resolution:="optional"
  Unresolved requirement: Import-Package: org.jboss.forge.roaster.model.source; resolution:="optional"
  Unresolved requirement: Require-Bundle: org.jboss.forge.roaster.roaster-api

	at org.eclipse.osgi.container.Module.start(Module.java:463)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1847)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor$1$1.execute(EquinoxContainerAdaptor.java:136)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1840)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1783)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1745)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1667)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:345)
```

It seems that roaster-jdt cannot find the roaster-api, however, it is included in the list of plugins. Why does this happen? Should I add anything else? This is the minimal example updated:

[minimal.example.project.zip](https://github.com/eclipse/tycho/files/8957964/minimal.example.project.zip)

Thanks. 

@laeubi Thanks for your answer. Not sure how to include the unit within feature. It is possible to do so? How?

Kind regards,
Antonio


[1] https://wiki.eclipse.org/Tycho/category.xml

**Reply by antoniogarmendia** [reply]
Hi all, 

I am still interested in solving this issue. At this moment, the only solution that I found is to write a bash that changes the auto start file in configuration\org.eclipse.equinox.simpleconfigurator\bundles.info, but of course, I am still looking for a way to automate this in order to provide a user-friendly installation process. If I need to improve the minimal example, just let me know.

Thanks.

Kind regards,
Antonio
